File: core/options/greeks/black_scholes.py
"""Vectorized Black-Scholes Greeks calculation.

Domain:    Options Analysis — Black-Scholes
Context:
  - Inputs come from yfinance option-chain snapshots.
  - Defensive against illiquid strikes (bid=0, IV=999.x).
  - European exercise assumption.
Contracts:
  - greeks_vectorized(S, K, T, r, sigma, option_type) -> dict[str, np.ndarray]
Dependencies UPWARD:
  - scipy.stats, numpy
Dependencies DOWNWARD:
  - core.options.greeks.portfolio, core.strategies
"""

from __future__ import annotations

import logging

import numpy as np
from scipy.stats import norm

logger = logging.getLogger(__name__)

_T_MIN = 1 / 365

# CONSTRAINT: zero-IV inputs would trigger division-by-zero in the d1 / d2 formulas.
_SIGMA_MIN = 0.001

# CONSTRAINT: corrupt chain data (e.g. IV=999) would explode exponent terms and swamp valid Greeks with NaNs.
_SIGMA_MAX = 20.0


def _safe_inputs(S, K, T, sigma):
    S = np.asarray(S, dtype=float)
    K = np.asarray(K, dtype=float)
    T = np.asarray(T, dtype=float)
    sigma = np.asarray(sigma, dtype=float)
    valid = (
        np.isfinite(S) & (S > 0) & np.isfinite(K) & (K > 0)
        & np.isfinite(sigma) & (sigma >= _SIGMA_MIN) & (sigma <= _SIGMA_MAX)
        & np.isfinite(T) & (T >= _T_MIN)
    )
    S_ = np.where(valid, S, 100.0)
    K_ = np.where(valid, K, 100.0)
    T_ = np.where(valid, T, 1 / 365)
    sigma_ = np.where(valid, sigma, 0.2)
    return S_, K_, T_, sigma_, valid


def greeks_vectorized(S, K, T, r, sigma, option_type="call"):
    """Vectorized Black-Scholes Greeks.

    Parameters can be scalars or equal-length NumPy arrays.
    Returns dict with same-shape arrays; invalid positions contain np.nan.
    """
    S_, K_, T_, sigma_, valid = _safe_inputs(S, K, T, sigma)
    r_ = float(r)
    sqrt_T = np.sqrt(T_)
    d1 = (np.log(S_ / K_) + (r_ + 0.5 * sigma_**2) * T_) / (sigma_ * sqrt_T)
    d2 = d1 - sigma_ * sqrt_T
    n_d1 = norm.pdf(d1)
    N_d1 = norm.cdf(d1)
    N_d2 = norm.cdf(d2)
    disc = np.exp(-r_ * T_)

    gamma = np.where(valid, n_d1 / (S_ * sigma_ * sqrt_T), np.nan)
    vega = np.where(valid, S_ * n_d1 * sqrt_T / 100, np.nan)

    if option_type == "call":
        delta = np.where(valid, N_d1, np.nan)
        theta = np.where(valid, (-(S_ * n_d1 * sigma_) / (2 * sqrt_T) - r_ * K_ * disc * N_d2) / 365, np.nan)
        price = np.where(valid, S_ * N_d1 - K_ * disc * N_d2, np.nan)
    else:
        delta = np.where(valid, N_d1 - 1, np.nan)
        theta = np.where(valid, (-(S_ * n_d1 * sigma_) / (2 * sqrt_T) + r_ * K_ * disc * norm.cdf(-d2)) / 365, np.nan)
        price = np.where(valid, K_ * disc * norm.cdf(-d2) - S_ * norm.cdf(-d1), np.nan)

    S_raw = np.asarray(S, dtype=float)
    K_raw = np.asarray(K, dtype=float)
    intrinsic = np.where(option_type == "call", np.maximum(S_raw - K_raw, 0), np.maximum(K_raw - S_raw, 0))

    return {
        "delta": delta,
        "gamma": gamma,
        "theta": theta,
        "vega": vega,
        "bs_price": price,
        "intrinsic": intrinsic,
        "time_value": np.where(valid, np.maximum(price - intrinsic, 0), np.nan),
    }

File: static/sim/black_scholes.js
// black_scholes.js — scalar Black–Scholes price + Greeks.
//
// Port of core/options/greeks/black_scholes.py::greeks_vectorized, kept scalar
// (the simulator evaluates at the current spot). The safety clamps are copied
// VERBATIM from the Python source — they are intentional domain constants
// (see docs/constraints.md §5), not magic numbers.

import { normCdf, normPdf } from './norm.js';

export const T_MIN = 1 / 365;
export const SIGMA_MIN = 0.001;
export const SIGMA_MAX = 20.0;

function _safeInputs(S, K, T, sigma) {
  const valid =
    Number.isFinite(S) && S > 0 &&
    Number.isFinite(K) && K > 0 &&
    Number.isFinite(sigma) && sigma >= SIGMA_MIN && sigma <= SIGMA_MAX &&
    Number.isFinite(T) && T >= T_MIN;
  return {
    valid,
    S: valid ? S : 100,
    K: valid ? K : 100,
    T: valid ? T : 1 / 365,
    sigma: valid ? sigma : 0.2,
  };
}

// Returns the same keys as the Python greeks_vectorized.
export function bsGreeks(S, K, T, r, sigma, optionType = 'call') {
  const { valid, S: S_, K: K_, T: T_, sigma: sigma_ } = _safeInputs(S, K, T, sigma);
  const sqrtT = Math.sqrt(T_);
  const d1 = (Math.log(S_ / K_) + (r + 0.5 * sigma_ * sigma_) * T_) / (sigma_ * sqrtT);
  const d2 = d1 - sigma_ * sqrtT;
  const n_d1 = normPdf(d1);
  const N_d1 = normCdf(d1);
  const N_d2 = normCdf(d2);
  const N_m_d2 = normCdf(-d2);
  const N_m_d1 = normCdf(-d1);
  const disc = Math.exp(-r * T_);

  const gamma = valid ? n_d1 / (S_ * sigma_ * sqrtT) : NaN;
  const vega = valid ? (S_ * n_d1 * sqrtT) / 100 : NaN;

  let delta, theta, price;
  if (optionType === 'call') {
    delta = valid ? N_d1 : NaN;
    theta = valid ? (-(S_ * n_d1 * sigma_) / (2 * sqrtT) - r * K_ * disc * N_d2) / 365 : NaN;
    price = valid ? S_ * N_d1 - K_ * disc * N_d2 : NaN;
  } else {
    delta = valid ? N_d1 - 1 : NaN;
    theta = valid ? (-(S_ * n_d1 * sigma_) / (2 * sqrtT) + r * K_ * disc * N_m_d2) / 365 : NaN;
    price = valid ? K_ * disc * N_m_d2 - S_ * N_m_d1 : NaN;
  }

  const intrinsic = optionType === 'call' ? Math.max(S - K, 0) : Math.max(K - S, 0);
  const timeValue = valid ? Math.max(price - intrinsic, 0) : NaN;

  return {
    delta,
    gamma,
    theta,
    vega,
    bs_price: price,
    intrinsic,
    time_value: timeValue,
  };
}
